execution/distinct.go

```diff
@@ -24,104 +28,162 @@ func (node *Distinct) Get(ctx context.Context, variables octosql.Variables, stre
+		newRecordCount = recordCount.AsInt() - 1
+	}
 
-func (ds *DistinctStream) Next(ctx context.Context) (*Record, error) {
-	for {
-		record, err := ds.stream.Next(ctx)
+	if newRecordCount != 0 {
+		newRecordCountValue := octosql.MakeInt(newRecordCount)
+		err := recordCountState.Set(&newRecordCountValue)
 		if err != nil {
-			if err == ErrEndOfStream {
-				return nil, ErrEndOfStream
-			}
-			return nil, errors.Wrap(err, "couldn't get record from stream in DistinctStream")
+			return errors.Wrap(err, "couldn't save record count")
 		}
-
-		already, err := ds.records.Has(record)
-
+	} else {
+		err := recordCountState.Clear()
 		if err != nil {
-			return nil, errors.Wrap(err, "couldn't access the record set")
+			return errors.Wrap(err, "couldn't clear record count")
 		}
+	}
 
-		if !already {
-			_, err := ds.records.Insert(record)
+	return nil
+}
 
-			if err != nil {
-				return nil, errors.Wrap(err, "couldn't access the record set")
-			}
+func (ds *DistinctStream) Trigger(ctx context.Context, tx storage.StateTransaction, key octosql.Value) ([]*Record, error) {
+	keyPrefix := append(append([]byte("$"), key.MonotonicMarshal()...), '$')
+	txByKey := tx.WithPrefix(keyPrefix)
 
-			return record, nil
-		}
-	}
-}
+	phantom := octosql.MakePhantom()
 
-type recordSet struct {
-	set map[uint64][]*Record
-}
+	// Check whether record was triggered
+	recordTriggerState := storage.NewValueState(txByKey.WithPrefix(previouslyTriggeredValuePrefix))
+	var wasRecordTriggered bool
 
-func newRecordSet() *recordSet {
-	return &recordSet{
-		set: make(map[uint64][]*Record),
+	err := recordTriggerState.Get(&phantom)
+	if err == nil {
+		wasRecordTriggered = true
+	} else if err == storage.ErrNotFound {
+		wasRecordTriggered = false
+	} else {
+		return nil, errors.Wrap(err, "couldn't check whether record was already triggered")
 	}
-}
 
-func (rs *recordSet) Has(r *Record) (bool, error) {
-	hash, err := r.Hash()
-	if err != nil {
-		return false, errors.Wrap(err, "couldn't get hash of record")
+	keyTuple := key.AsSlice()
+	if len(keyTuple) == 0 { // empty record
+		return nil, nil
+	}
+
+	recordMap := keyTuple[0].AsMap()
+
+	// Check whether record is present
+	recordCountState := storage.NewValueState(txByKey.WithPrefix(recordCountPrefix))
+	var isRecordPresent bool
+
+	err = recordCountState.Get(&phantom)
+	if err == nil {
+		isRecordPresent = true
+	} else if err == storage.ErrNotFound {
+		isRecordPresent = false
+	} else {
+		return nil, errors.Wrap(err, "couldn't check whether record is present in state")
 	}
 
-	for i := range rs.set[hash] {
-		if r.Equal(rs.set[hash][i]) {
-			return true, nil
+	if !wasRecordTriggered && isRecordPresent {
+		// This is the first time that record is triggered and it is present, so we send it
+
+		// Save that the record was triggered
+		err = recordTriggerState.Set(&phantom)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't mark record as retracted")
+		}
+
+		record := recordFromMap(recordMap, false)
```

Comment:
Maybe just store the record somewhere instead of reconstructing it from the key.

This way we will surely not lose stuff like metadata, event time, an ID etc.